File: domain/src/test/java/cn/byronlab/weather/domain/usecase/SearchCitiesUseCaseTest.java
package cn.byronlab.weather.domain.usecase;

import cn.byronlab.weather.domain.model.City;
import cn.byronlab.weather.domain.model.Unit;
import cn.byronlab.weather.domain.repository.CityRepository;
import cn.byronlab.weather.domain.result.DomainResult;

import org.junit.Test;

import java.util.Collections;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

public class SearchCitiesUseCaseTest {

    @Test
    public void blankKeywordReturnsEmptyListWithoutCallingRepository() {
        FakeCityRepository repository = new FakeCityRepository();
        SearchCitiesUseCase useCase = new SearchCitiesUseCase(repository);

        DomainResult<List<City>> result = useCase.execute("  ");

        assertTrue(result.isSuccess());
        assertTrue(result.getData().isEmpty());
        assertNull(repository.lastSearchKeyword);
    }

    @Test
    public void nonBlankKeywordIsTrimmedBeforeSearch() {
        FakeCityRepository repository = new FakeCityRepository();
        SearchCitiesUseCase useCase = new SearchCitiesUseCase(repository);

        DomainResult<List<City>> result = useCase.execute(" 北京 ");

        assertTrue(result.isSuccess());
        assertEquals("北京", repository.lastSearchKeyword);
    }

    private static final class FakeCityRepository implements CityRepository {

        private String lastSearchKeyword;

        @Override
        public DomainResult<List<City>> searchCities(String keyword) {
            lastSearchKeyword = keyword;
            return DomainResult.success(Collections.singletonList(
                    new City("101010100", "北京", "beijing", "", "", "116.4", "39.9")
            ));
        }

        @Override
        public DomainResult<List<City>> getSavedCities() {
            return DomainResult.success(Collections.emptyList());
        }

        @Override
        public DomainResult<Unit> deleteSavedCity(String cityId) {
            return DomainResult.success(Unit.INSTANCE);
        }
    }
}

File: domain/src/test/java/cn/byronlab/weather/domain/usecase/SearchCitiesUseCaseTest.kt
package cn.byronlab.weather.domain.usecase

import cn.byronlab.weather.domain.model.City
import cn.byronlab.weather.domain.model.DeviceLocation
import cn.byronlab.weather.domain.repository.CityRepository
import cn.byronlab.weather.domain.result.DomainResult
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class SearchCitiesUseCaseTest {

    @Test
    fun blankKeywordReturnsEmptyListWithoutCallingRepository() = runTest {
        val repository = FakeCityRepository()
        val useCase = SearchCitiesUseCase(repository)

        val result = useCase("  ")

        assertTrue(result is DomainResult.Success)
        assertTrue((result as DomainResult.Success).data.isEmpty())
        assertNull(repository.lastSearchKeyword)
    }

    @Test
    fun nonBlankKeywordIsTrimmedBeforeSearch() = runTest {
        val repository = FakeCityRepository()
        val useCase = SearchCitiesUseCase(repository)

        val result = useCase(" 北京 ")

        assertTrue(result is DomainResult.Success)
        assertEquals("北京", repository.lastSearchKeyword)
    }

    private class FakeCityRepository : CityRepository {

        var lastSearchKeyword: String? = null

        override suspend fun searchCities(keyword: String): DomainResult<List<City>> {
            lastSearchKeyword = keyword
            return DomainResult.success(listOf(City("101010100", "北京", "beijing", "", "", "116.4", "39.9")))
        }

        override suspend fun getPopularCities(): DomainResult<List<City>> {
            return DomainResult.success(emptyList())
        }

        override fun resolveLocation(location: DeviceLocation): DomainResult<City> {
            return DomainResult.success(City("location", location.cityName, location.cityName, "", "", "", ""))
        }

        override fun observeAddedCities(): Flow<DomainResult<List<City>>> = flow {
            emit(getAddedCities())
        }

        override suspend fun getAddedCities(): DomainResult<List<City>> {
            return DomainResult.success(emptyList())
        }

        override suspend fun addCity(cityId: String): DomainResult<Unit> {
            return DomainResult.success(Unit)
        }

        override suspend fun removeCity(cityId: String): DomainResult<Unit> {
            return DomainResult.success(Unit)
        }

        override suspend fun getRecentCities() = DomainResult.success(emptyList<City>())

        override suspend fun recordRecentCity(cityId: String) = DomainResult.success(Unit)
    }
}
